crates/shadowsocks-service/src/acl/mod.rs

```diff
@@ -158,25 +161,76 @@ impl ParsingRules {
     }
 
     fn add_ipv4_rule(&mut self, rule: impl Into<Ipv4Net>) {
-        self.ipv4.add(rule.into());
+        let rule = rule.into();
+        trace!("IPV4-RULE {}", rule);
+        self.ipv4.add(rule);
     }
 
     fn add_ipv6_rule(&mut self, rule: impl Into<Ipv6Net>) {
-        self.ipv6.add(rule.into());
+        let rule = rule.into();
+        trace!("IPV6-RULE {}", rule);
+        self.ipv6.add(rule);
     }
 
     fn add_regex_rule(&mut self, mut rule: String) {
+        static TREE_SET_RULE_EQUIV: Lazy<Regex> = Lazy::new(|| {
+            RegexBuilder::new(
+                r#"^(?:(?:\((?:\?:)?\^\|\\\.\)|(?:\^\.\+)?\\\.)((?:[\w-]+(?:\\\.)?)+)|\^((?:[\w-]+(?:\\\.)?)+))\$$"#,
+            )
+            .unicode(false)
+            .build()
+            .unwrap()
+        });
+
+        if let Some(caps) = TREE_SET_RULE_EQUIV.captures(rule.as_bytes()) {
+            if let Some(tree_rule) = caps.get(1) {
+                if let Ok(tree_rule) = str::from_utf8(tree_rule.as_bytes()) {
+                    let tree_rule = tree_rule.replace("\\.", ".");
+                    if let Ok(..) = self.add_tree_rule_inner(&tree_rule) {
+                        trace!("REGEX-RULE {} => TREE-RULE {}", rule, tree_rule);
+                        return;
+                    }
+                }
+            }
+
+            if let Some(set_rule) = caps.get(2) {
```

Comment:
```suggestion
            } else if let Some(set_rule) = caps.get(2) {
```
Since it's not possible for both rules match at the same time you can use else if here.